spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+  if !isWithinTrustedPeriod(h2) { return ErrHeaderNotWithinTrustedPeriod(h2) }
 
-_Remark_: The variable _trustlevel_ can be used if the user believes that relying on one correct validator is not sufficient. However, in case of (frequent) changes in the validator set, the higher the _trustlevel_ is chosen, the more unlikely it becomes that CheckSupport returns true for non-adjacent headers.
+  // get the highest trusted headers lower than h2
+  h1 = Store.HighestTrustedSmallerThan(height)
+  if h1 == nil { return ErrNoTrustedHeader }
 
-**Bisection.** The following function uses CheckSupport in a recursion to find intermediate headers that allow to establish a sequence of trust.
+  err = CanTrust(h1, h2, trustThreshold)  // or CanTrustBisection((h1, h2, trustThreshold)
+  if err != nil { return err }
 
-```go
-func Bisection(h1,h2,trustlevel) bool{
-  if CheckSupport(h1,h2,trustlevel) {
-    return true
+  if isWithinTrustedPeriod(h2) {
+    Store.add(h2)
+    // we store only trusted headers, as we assume that only trusted headers
+    // are influencing end user business decisions.
+    return nil
   }
-  if h2.Header.height == h1.Header.height + 1 {
-    // we have adjacent headers that are not matching (failed
-    // the CheckSupport)
-    // we could submit evidence here
-    return false
+  return ErrHeaderNotTrusted(h2)
+}
+
+
+// return nil in case we can trust header h2 based on header h1; otherwise return error
+// where error captures the nature of the error.
+func CanTrust(h1,h2,trustThreshold) error {
+  assume h1.Header.Height < h2.header.Height
+
+  th := h1 // th is trusted header
+  untrustedHeaders := [h2]
+
+  while true {
+    for h in untrustedHeaders {
+        // we assume here that iteration is done in the order of header heights
+        err = CheckSupport(th,h,trustThreshold)
+        if err == nil {
+            th = h
+            Store.Add(h)
+            untrustedHeaders.RemoveHeadersSmallerOrEqual(h.Header.Height)
+            if th == h2 { return nil }
+        }
+        if (err != ErrTooMuchChange) { return err }
+    }
+
+    endHeight = min(untrustedHeaders)
+    foundPivot = false
+    while(!foundPivot) {
+        pivot := (th.Header.height + endHeight) / 2
+        hp := Commit(pivot)
+        if !verify(hp) { return ErrInvalidHeader(hp) }
+        // try to move trusted header forward to hp
+        err = CheckSupport(th,hp,trustThreshold)
+        if (err != nil and err != ErrTooMuchChange) return err
+        if err == nil {
+            th = hp
+            Store.Add(hp)
+            foundPivot = true
+        }
+        untrustedHeaders.add(hp)
+        endHeight = pivot
+    }
   }
+  return nil // this line should never be reached
+}
+```
+
+```go
+func CanTrustBisection(h1,h2,trustThreshold) error {
+  assume h1.Header.Height < h2.header.Height
+
+  err = CheckSupport(h1,h2,trustThreshold)
+  if err == nil {
+    Store.Add(h2)
+    return nil
+  }
+  if err != ErrTooMuchChange return err
```

Comment:
If we do this here, doesn't it mean we never start bisecting?

Reply:
We are in line 359 with `err != nil` and, if `err !=  ErrTooMuchChange`, the only possible values for `err` (looking at `CheckSupport()`) are:
  - `ErrHeaderNotWithinTrustedPeriod`
  - `ErrInvalidAdjacentHeaders`

They are both indicators that current bisection has failed, further recursive calls should not be done and therefore the call stack will unwind propagating this error to the final caller (`VerifyHeader()`).

I mentioned this in an earlier comment, it is better to write 359 and others like:
`if fatalCheckSupportError(err) return err`
Also, potential future error types added to `CheckSupport()` would be easier to implement without changing all the callers.

Reply:
Alternatively, we could use sth else than an error to indicate that there was too much change in the validator set. Returning a bool for instance. Or simply renaming the method name to indicate the only relevant case directly: `ValidatorsChangedTooMuch` or `EnoughIntersection` (I'm sure there are better names)

FWIW, linking the related discussion regarding the rust implementation: https://github.com/interchainio/tendermint-rs/pull/100/files#r360502272